src/thread_task.rs

```diff
@@ -301,68 +90,60 @@ impl Rav1dTasks {
+                    //  2. thread A begins `merge_pending_frame`, and clears `pending_tasks_merge`.
+                    //  3. thread B goes to add a pending task and takes the vector lock.
+                    //  4. thread A waits on the vector lock.
+                    //  5. thread B inserts some tasks, releases the lock, and sets `pending_tasks_merge`.
+                    //  6. thread A can now take the lock, and drain the vector.
+                    // We now have an empty vector: A has drained both the original contents,
+                    // and the task added by thread B - but `pending_tasks_merge` is true,
+                    // because it was set by B.
+                    //
+                    // So, if `merge_pending_frame` is called before another pending task is added,
+                    // we will get this spurious merge.
+                    return false;
                 }
-                let start = tasks.len() as u32;
                 tasks.extend(pending_tasks.drain(..));
-                start..tasks.len() as u32
-            };
-
-            for i in range {
-                // 1-based index, so we have to add 1
-                let task_idx = Rav1dTaskIndex(NonZeroU32::new(i + 1));
-                self.insert_tasks(c, task_idx, task_idx, 0);
             }
-        }
-        merge
-    }
-}
-
-impl Rav1dFrameContextTaskThread {
-    fn insert_task(&self, c: &Rav1dContext, task: Rav1dTask, cond_signal: c_int) -> Rav1dTaskIndex {
-        let idx = self.tasks.push(task);
-        self.tasks.insert_tasks(c, idx, idx, cond_signal);
-        idx
-    }
-}
-
-/// 1-based index into the task queue vector. 0 is reserved for None.
-#[derive(Copy, Clone, Default, PartialEq, Eq, PartialOrd, Ord, Debug, Atom)]
-#[repr(transparent)]
-pub struct Rav1dTaskIndex(Option<NonZeroU32>);
-
-impl Rav1dTaskIndex {
-    pub const NONE: Self = Self(None);
-
-    // Return the zero-based index into the task queue vector or `None`
-    pub fn raw_index(self) -> Option<u32> {
-        self.0.map(|i| i.get() - 1)
-    }
 
-    pub fn is_none(self) -> bool {
-        self.0.is_none()
-    }
+            // Unlike in the C code, we only do a single `insert_tasks`. There are two reasons:
+            //
+            //  - in the C code, the `insert_tasks` is what creates the sorted linked list.
+            //    In the Rust code, the sorting is done via sorting the `Vec` explictly.
+            //
+            //  - The key remaining function of is updating our position in the frame contexts,
+            //    by updating the `reset_task_cur` and `cur` variables. Here, as in C,
+            //    every task we're merging is for the same frame context, so the arguments are always the same,
+            //    and nothing is changed by trying to repeatedly do the same update.
+            let frame_idx = tasks[new_start].frame_idx;
+            self.insert_tasks(c, frame_idx, 0);
+            tasks.sort();
```

Comment:
I'm not sure, but is a stable sort needed here?  I would think it isn't, as `fn cmp` doesn't even ever return `Equal`.